overlord/configstate/config/transaction.go

```diff
@@ -103,6 +104,47 @@ func (t *Transaction) Changes() []string {
 	return out
 }
 
+// shadowsVirtualConfig checks that the given subkeys/value does not
+// "block" the path to a virtual config with a non-map type. E.g. if
+// "network.netplan" is virtual it must be impossible to set
+// "network=false" or getting the document under "network" would be
+// wrong.
+func shadowsVirtualConfig(instanceName string, subkeys []string, value interface{}) error {
+	// maps never block the path
+	if v := reflect.ValueOf(value); v.Kind() == reflect.Map {
+		return nil
+	}
+
+	// helper to compare if two subkeys slices have a common prefix
+	keysHaveCommonPrefix := func(s, prefix []string) bool {
+		if len(prefix) > len(s) {
+			return false
+		}
+		for i := range prefix {
+			if prefix[i] != s[i] {
+				return false
+			}
+		}
+		return true
+	}
+
+	virtualMu.Lock()
+	km := virtualMap[instanceName]
+	virtualMu.Unlock()
+
+	for virtualKey := range km {
+		virtualSubKeys, err := ParseKey(virtualKey)
+		if err != nil {
+			return fmt.Errorf("internal error: invalid virtual configuration: %v", err)
+		}
+		if len(virtualSubKeys) > len(subkeys) && keysHaveCommonPrefix(virtualSubKeys, subkeys) {
```

Comment:
I find it slightly confusing the we do the > comparison here, but then one is done also inside the helper as well,  what we don't want is emit the error if len(virtualSubKey) == len(subkeys) right?  so maybe here we should just check that they are different and let the helper do the rest?

Reply:
I like that, thanks - that makes it clearer indeed to just do `if len(virtualSubKeys) != len(subkeys) && keysHaveCommonPrefix(...`